### ISSUE TITLE
Fields are seemingly references by their position in an array

### ISSUE THREAD

**Issue by meldiner** [metabase/metabase#40046]
### Describe the bug

When doing operations post-summarize, fields are seemingly references by their position in an array. This means that if you delete fields that were summarized, all of the fields that you reference beneath the summary are now potentially offset and wrong.

### To Reproduce

![image](https://github.com/metabase/metabase/assets/9262738/a722930b-2450-43ba-8d8f-ed37290dce1a)
`board_name` custom field is set to [Max of Board Name]
I deleted the `Max of Task Status` field from the summarization section
![image](https://github.com/metabase/metabase/assets/9262738/5e723364-d54d-4179-b5e2-1af6f574d5af)
now it's set to [max_3]
which doesn't reconcile
![image](https://github.com/metabase/metabase/assets/9262738/ffe8363f-37a0-463a-840a-f77241123b30)
and things that referenced [Max of Status Changed at] now reference [Max of Board Name]
Things that referenced [Max of Task Status] now reference [Max of Status Changed At] 
basically, instead of referencing a field name or some unique identifier, they're referencing by position in the summarized fields array
and so when you delete one, everything equal or higher in the array becomes offset

### Expected behavior

References should point to the right fields

### Information about your Metabase installation

```JSON
Metabase v46.2
```

### Severity

annoying

**Comment by bshepherdson** [comment]
You're exactly right about what's going on here. In legacy MBQL, aggregation references are done by an index into the aggregations list.

MLv2 (new library and MBQL format) use UUIDs to reference aggregations, avoiding this problem. But legacy is still the source of truth in the appdb and on the wire, so we're not free of this bug yet.

**Comment by meldiner** [comment]
@bshepherdson, thanks for confirming. Do you expect updating metabase version to resolve this?

**Comment by bshepherdson** [comment]
I regret to say "not yet". I just checked this with the 0.49 RC and while the behavior has changed slightly (the expression is less broken after making the edit), this has not been fixed.

This is one of the diverse symptoms of a fundamental issue #36185 with how Metabase handles similar but not identical columns. There are lots of other ways to trip over this bug, such as joining the same table twice, or naming a custom column the same as a column from an earlier stage.

When the first aggregation `max` is deleted, the two surviving ones are renamed from `max_2` and `max_3` to `max` and `max_2`. The expression in the second stage of the query has a reference like `[:field "max_2"]`, which is still valid. We don't currently connect the dots that the ~artist~ column formerly known as `max_2` is now known as `max`.

So this won't be fixed in 49 or 50, but fixing that issue is one of my long-term goals. Much of the groundwork for it has landed over the past six months or so, but we're not ready to eliminate this issue yet.

**Comment by meldiner** [comment]
Hello, @bshepherdson. Our customers are repeatedly encountering this issue, which makes the data in Metabase perceived as unreliable. This is a high-priority bug for us to fix.

**Comment by NevRA** [comment]
@lbrdnk Braden will handle this

**Comment by meldiner** [comment]
@lbrdnk, it looks like a fix for this one was merged. That's great!

How can I check the Metabase version where the fix will be available?

**Comment by lbrdnk** [comment]
@meldiner, backport was merged, hence the milestone was added to the issue. That's the version in which the fix will be present.
